File: Rate Monotonic/RM.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

// Kiểm tra tính khả thi của hệ thống theo RM
bool is_schedulable(const vector<pair<int, int>>& tasks)
{
    int n = tasks.size();
    double U = 0;

    for (auto& task : tasks) {
        U += (double)task.first / task.second;
    }

    double bound = n * (pow(2.0, 1.0 / n) - 1);
    cout << "\nTổng tải hệ thống (U) = " << U << endl;
    cout << "Ngưỡng cho phép theo RM = " << bound << endl;

    return U <= bound;
}

// Hàm lập lịch Rate Monotonic
vector<int> rm_schedule(const vector<pair<int, int>>& tasks, int duration = -1) {
    int n = tasks.size();
    vector<pair<int, int>> sorted_tasks = tasks;

    // Sắp xếp theo chu kỳ tăng dần
    sort(sorted_tasks.begin(), sorted_tasks.end(), [](const pair<int, int>& a, const pair<int, int>& b)
    {
        return a.second < b.second;
    });

    if (duration == -1) {
        int max_period = 0;
        for (auto& task : sorted_tasks) {
            if (task.second > max_period)
                max_period = task.second;
        }
        duration = 3 * max_period;
    }

    vector<int> timeline(duration, 0);
    vector<int> remaining_time(n, 0);
    vector<int> release_time(n, 0);

    for (int t = 0; t < duration; ++t)
    {
        // Cập nhật các task được phát hành tại thời điểm t
        for (int i = 0; i < n; ++i)
        {
            if (t % sorted_tasks[i].second == 0)
            {
                remaining_time[i] = sorted_tasks[i].first;
                release_time[i] = t;
            }
        }

        // Chọn task có độ ưu tiên cao nhất (chu kỳ nhỏ nhất) còn thời gian
        for (int i = 0; i < n; ++i)
        {
            if (remaining_time[i] > 0 && release_time[i] <= t)
            {
                timeline[t] = i + 1; // Task i+1 được thực hiện tại thời điểm t
                remaining_time[i]--;
                break;
            }
        }
    }
    return timeline;
}

// --- Hàm main ---
int main()
{
    int n;
    cout << "Nhập số lượng tasks: ";
    cin >> n;

    vector<pair<int, int>> tasks(n);
    cout << "Nhập thời gian thực thi và chu kỳ của mỗi task (C T):\n";
    for (int i = 0; i < n; ++i)
    {
        cout << "Task " << i + 1 << ": ";
        cin >> tasks[i].first >> tasks[i].second;
    }

    // Kiểm tra schedulability
    if (!is_schedulable(tasks))
    {
        cout << "\n!!! Hệ thống KHÔNG đảm bảo thực hiện đúng hạn theo Rate Monotonic.\n";
    }
    else
    {
        cout << "\n>>> Hệ thống CÓ THỂ thực hiện đúng hạn theo Rate Monotonic.\n";
    }

    // Thực hiện lập lịch
    vector<int> timeline = rm_schedule(tasks);

    // In kết quả lập lịch
    cout << "\nLịch thực hiện các task theo từng giây:\n";
    for (size_t i = 0; i < timeline.size(); ++i)
    {
        cout << timeline[i];
        if (i != timeline.size() - 1) cout << ", ";
    }
    cout << endl;

    return 0;
}

File: Rate Monotonic/RM.py
# Lập lịch RM
def rm_schedule(tasks, duration=None):
    # Sắp xếp tasks theo chu kỳ tăng dần (ưu tiên cao nhất là task có chu kỳ nhỏ nhất)
    tasks = sorted(tasks, key=lambda x: x[1])

    if duration is None:
        max_period = max(t[1] for t in tasks)
        duration = 3 * max_period

    n = len(tasks)
    timeline = [0] * duration
    remaining_time = [0] * n
    release_time = [0] * n  # Thời điểm phát hành gần nhất cho từng task

    for t in range(duration):
        # Cập nhật các task được phát hành tại thời điểm t
        for i in range(n):
            if t % tasks[i][1] == 0:
                remaining_time[i] = tasks[i][0]
                release_time[i] = t

        # Chọn task ưu tiên cao nhất có remaining_time > 0
        for i in range(n):
            if remaining_time[i] > 0 and release_time[i] <= t:
                timeline[t] = i + 1  # Đánh dấu task chạy tại thời điểm t
                remaining_time[i] -= 1
                break  # Task ưu tiên cao nhất được chạy, ngắt các task thấp hơn

    return timeline, tasks


# Vẽ biểu đồ
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches

def draw_timing_diagram(timeline, task_list):
    fig, ax = plt.subplots(figsize=(16, 6))
    n_tasks = len(task_list)
    colors = ['skyblue', 'lightgreen', 'salmon', 'orange', 'violet']
    end_time = len(timeline)

    # Vẽ mỗi task một dòng
    for i in range(n_tasks):
        y = n_tasks - i - 1  # để dòng trên cùng là Task 1
        start = None
        for t in range(end_time):
            if timeline[t] == i + 1:
                if start is None:
                    start = t
            else:
                if start is not None:
                    ax.broken_barh([(start, t - start)], (y - 0.4, 0.8),
                                   facecolors=colors[i % len(colors)])
                    for tt in range(start, t):
                        ax.text(tt + 0.2, y, str(tt), fontsize=8)
                    start = None
        if start is not None:
            ax.broken_barh([(start, end_time - start)], (y - 0.4, 0.8),
                           facecolors=colors[i % len(colors)])
            for tt in range(start, end_time):
                ax.text(tt + 0.2, y, str(tt), fontsize=8)

    # Vẽ đường phân cách chu kỳ riêng cho từng task
            
    for idx, task in enumerate(task_list):
        _, T = task
        # Đảo chiều để vẽ từ dòng cao xuống thấp
        reversed_idx = n_tasks - idx - 1
        for p in range(0, end_time + 1, T):
            ax.axvline(
                x=p,
                ymin=(reversed_idx) / n_tasks,
                ymax=(reversed_idx + 1) / n_tasks,
                color='black', linestyle='--', linewidth=1.5
            )

    # Cấu hình trục và nhãn
    ax.set_yticks(range(n_tasks))
    ax.set_yticklabels([f'Task {n_tasks - i}' for i in (range(n_tasks))], fontsize = 20)
    ax.set_xticks(range(0, end_time + 1))  # giây tròn
    ax.set_xticklabels([str(i) for i in range(0, end_time + 1)], fontsize = 12)
    ax.set_xlim(0, end_time)
    ax.set_xlabel("Time (seconds)", fontsize = 15)
    ax.set_title("RATE MONOTONIC SCHEDULING DIAGRAM", fontsize = 30)
    plt.grid(axis='x', linestyle=':', linewidth=0.5)

    # Thêm chú thích màu task
    legend = [mpatches.Patch(color=colors[i % len(colors)], label=f'Task {i + 1}') for i in range(n_tasks)]
    ax.legend(handles=legend)
    plt.tight_layout()
    plt.show()

# --- Ví dụ chạy thử ---
if __name__ == "__main__":
    # Mỗi task: (thời gian thực thi, chu kỳ)
    tasks = [(3,20), (2,5), (2,10)]  # Task 1, 2, 3

    timeline, tasks = rm_schedule(tasks)
    draw_timing_diagram(timeline, tasks)
